src/types/quota.rs

```diff
@@ -0,0 +1,233 @@
+    }
+
+    /// The defined resources with their usage and limits (could be empty)
+    pub fn resources(&self) -> &[QuotaResource<'a>] {
+        &self.resources[..]
+    }
+}
+
+impl<'a> Quota<'a> {
+    fn from_imap_proto(q: imap_proto::Quota<'a>) -> Self {
+        Self {
+            root_name: q.root_name,
+            resources: q
+                .resources
+                .into_iter()
+                .map(|e| QuotaResource {
+                    name: match e.name {
+                        imap_proto::QuotaResourceName::Storage => QuotaResourceName::Storage,
+                        imap_proto::QuotaResourceName::Message => QuotaResourceName::Message,
+                        imap_proto::QuotaResourceName::Atom(e) => QuotaResourceName::Atom(e),
+                    },
+                    usage: e.usage,
+                    limit: e.limit,
+                })
+                .collect(),
+        }
+    }
+}
+
+/// From [QUOTA Response](https://datatracker.ietf.org/doc/html/rfc2087#section-5.1)
+///
+/// The quota resource sub pieces int he Quota response. Used by [Quota]
+#[derive(Debug, Eq, PartialEq)]
+pub struct QuotaResource<'a> {
+    /// The resource type
+    pub name: QuotaResourceName<'a>,
+    /// current usage of the resource
+    pub usage: u64,
+    /// resource limit
+    pub limit: u64,
+}
+
+/// From [QUOTAROOT Response](https://datatracker.ietf.org/doc/html/rfc2087#section-5.2)
+///
+/// Used by [Session::get_quota_root]
+#[self_referencing]
+pub struct QuotaRoot {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) inner: InnerQuotaRoot<'this>,
+}
+
+/// Inner struct to manage storing the references for ouroboros
+pub(crate) struct InnerQuotaRoot<'a> {
+    pub(crate) quota_root: imap_proto::QuotaRoot<'a>,
+    pub(crate) quotas: Vec<Quota<'a>>,
+}
+
+impl QuotaRoot {
+    /// Parse the [`QuotaRoot`] response from a response buffer.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        QuotaRootTryBuilder {
+            data: owned,
+            inner_builder: |input| {
+                let mut quota_roots = Vec::new();
+                let mut quotas = Vec::new();
+
+                parse_many_into2(
+                    input,
+                    &mut quota_roots,
+                    &mut quotas,
+                    unsolicited,
+                    |response| match response {
+                        Response::QuotaRoot(q) => Ok(MapOrNot2::Map1(q)),
+                        Response::Quota(q) => Ok(MapOrNot2::Map2(Quota::from_imap_proto(q))),
+                        resp => Ok(MapOrNot2::Not(resp)),
+                    },
+                )?;
+
+                if quota_roots.is_empty() {
+                    Err(Error::Parse(ParseError::Invalid(input.to_vec())))
+                } else {
+                    Ok(InnerQuotaRoot {
+                        quota_root: quota_roots.remove(0),
```

Comment:
Should this error if there's >1?

Reply:
And also do something other than panic if there are none (also a question for if we implement a helper)?

Reply:
I reworked the handling here and in the Quota::parse to explicitly handle 1 vs not 1 and added tests to ensure (no panic) handling of those scenarios.